File: scraper_service/freshershunt_bypass.js
const puppeteer = require('puppeteer-extra');
const StealthPlugin = require('puppeteer-extra-plugin-stealth');
puppeteer.use(StealthPlugin());
const AdblockerPlugin = require('puppeteer-extra-plugin-adblocker');
puppeteer.use(AdblockerPlugin({ blockTrackers: true }));

async function scrapeFreshershunt(limit = 10) {
    const browser = await puppeteer.launch({ 
        headless: "new",
        args: ['--no-sandbox', '--disable-setuid-sandbox', '--disable-web-security']
    });
    const page = await browser.newPage();
    const jobs = [];

    try {
        await page.goto('https://www.freshershunt.com/', { waitUntil: 'domcontentloaded', timeout: 60000 });
        
        const postLinks = await page.evaluate(() => {
            const links = Array.from(document.querySelectorAll('h2 a, h3 a, article a'));
            return links.map(a => ({ title: a.innerText, url: a.href })).filter(j => j.url && j.title);
        });

        // Deduplicate
        const uniqueLinksMap = new Map();
        for (const post of postLinks) {
            if (!uniqueLinksMap.has(post.url)) {
                uniqueLinksMap.set(post.url, post);
            }
        }
        const uniqueLinks = Array.from(uniqueLinksMap.values()).slice(0, limit);

        for (const post of uniqueLinks) {
            const jobPage = await browser.newPage();
            try {
                await jobPage.goto(post.url, { waitUntil: 'domcontentloaded', timeout: 30000 });
                
                // Find "Apply Now" or "Click Here" link in the article body
                const applyLink = await jobPage.evaluate(() => {
                    const buttons = Array.from(document.querySelectorAll('a'));
                    const applyBtn = buttons.find(b => {
                        const t = b.innerText.toLowerCase();
                        return t.includes('apply') || t.includes('click here') || t.includes('registration link');
                    });
                    return applyBtn ? applyBtn.href : null;
                });
                
                if (applyLink) {
                    await jobPage.goto(applyLink, { waitUntil: 'domcontentloaded', timeout: 30000 });
                    
                    // Ad-gateway bypass logic
                    // If it redirects immediately, we just wait a bit
                    await new Promise(r => setTimeout(r, 6000));
                    
                    let finalUrl = await jobPage.evaluate(() => {
                        // Sometimes there's a skip button
                        const skipBtn = document.querySelector('.skip-btn, #skip, a.get-link, a#getlink');
                        if (skipBtn && skipBtn.href) return skipBtn.href;
                        return window.location.href; 
                    });
                    
                    // Basic sanity check - if the final URL is still freshershunt, it failed to extract
                    if (!finalUrl.includes('freshershunt.com')) {
                        jobs.push({
                            title: post.title,
                            source: 'freshershunt',
                            url: finalUrl,
                            original_post: post.url
                        });
                    }
                }
            } catch (err) {
                // Silently skip on error to keep processing
            } finally {
                await jobPage.close();
            }
        }
    } catch (err) {
        console.error('Failed to scrape Freshershunt:', err);
    } finally {
        await browser.close();
    }
    
    console.log(JSON.stringify(jobs));
}

scrapeFreshershunt(process.argv[2] ? parseInt(process.argv[2]) : 10);

File: scraper_service/freshershunt_bypass.py
import json
import logging
import sys
import time
from playwright.sync_api import sync_playwright
from playwright_stealth import Stealth

logging.basicConfig(level=logging.INFO, stream=sys.stderr)
logger = logging.getLogger(__name__)

def scrape_freshershunt(limit: int = 10) -> list:
    jobs = []
    stealth = Stealth()
    
    try:
        with sync_playwright() as p:
            with p.chromium.launch(
                headless=True,
                args=['--no-sandbox', '--disable-setuid-sandbox', '--disable-web-security']
            ) as browser:
                context = browser.new_context()
                page = context.new_page()
                stealth.apply_stealth_sync(page)
                
                # We can't trivially adblock like puppeteer-extra-plugin-adblocker, 
                # but Playwright's router can abort image/media requests if needed.
                page.route("**/*", lambda route: route.abort() if route.request.resource_type in ["image", "media", "font"] else route.continue_())
                
                page.goto('https://www.freshershunt.com/', wait_until='domcontentloaded', timeout=60000)
                
                post_links = page.evaluate("""() => {
                    const links = Array.from(document.querySelectorAll('h2 a, h3 a, article a'));
                    return links.map(a => ({ title: a.innerText, url: a.href })).filter(j => j.url && j.title);
                }""")
                
                unique_links = {}
                for post in post_links:
                    if post['url'] not in unique_links:
                        unique_links[post['url']] = post
                
                posts_to_process = list(unique_links.values())[:limit]
                
                for post in posts_to_process:
                    job_page = context.new_page()
                    stealth.apply_stealth_sync(job_page)
                    try:
                        job_page.goto(post['url'], wait_until='domcontentloaded', timeout=30000)
                        
                        apply_link = job_page.evaluate("""() => {
                            const buttons = Array.from(document.querySelectorAll('a'));
                            const applyBtn = buttons.find(b => {
                                const t = b.innerText.toLowerCase();
                                return t.includes('apply') || t.includes('click here') || t.includes('registration link');
                            });
                            return applyBtn ? applyBtn.href : null;
                        }""")
                        
                        if apply_link:
                            job_page.goto(apply_link, wait_until='domcontentloaded', timeout=30000)
                            time.sleep(6)
                            
                            final_url = job_page.evaluate("""() => {
                                const skipBtn = document.querySelector('.skip-btn, #skip, a.get-link, a#getlink');
                                if (skipBtn && skipBtn.href) return skipBtn.href;
                                return window.location.href; 
                            }""")
                            
                            if 'freshershunt.com' not in final_url:
                                jobs.append({
                                    "title": post['title'],
                                    "source": 'freshershunt',
                                    "url": final_url,
                                    "original_post": post['url']
                                })
                    except Exception:
                        pass
                    finally:
                        job_page.close()
    except Exception as e:
        logger.error(f"Failed to scrape Freshershunt: {e}")
        
    return jobs

if __name__ == "__main__":
    limit = int(sys.argv[1]) if len(sys.argv) > 1 else 10
    res = scrape_freshershunt(limit)
    print(json.dumps(res))
